src/Orleans/Streams/SimpleMessageStream/SimpleMessageStreamProducer.cs

```diff
@@ -86,7 +95,12 @@ internal class SimpleMessageStreamProducer<T> : IInternalAsyncBatchObserver<T>
             if (isDisposed) throw new ObjectDisposedException(string.Format("{0}-{1}", GetType(), "OnNextAsync"));
 
             if (!connectedToRendezvous)
+            {
+                // In order to avoid potential concurrency errors, synchronously copy the input before yielding the
+                // thread. DeliverItem below must also be take care to avoid yielding before copying for non-immutable objects.
+                item = (T)this.serializationManager.DeepCopy(item);
```

Comment:
I think optimizeForImmutableData is for configuring sending remote message to consumer grain, right? And this is before that step, the preparation step just to call its own prducerExtension to deliver this item, which is not something optimizeForImmutableData is designed to configure.